### PR TITLE
Fixes for test case failures after the merge to LLVM/Clang 12.x

### PR DESCRIPTION
**Pull request by sulekhark**
This PR shows the diffs of the branch `updated_baseline_master_12` with respect to the branch `updated_baseline_master_12_checkpt`. It is created only to enable the review of the fixes for the test failures due to merge, without cluttering the PR with LLVM 12.x changes. After the review, the branch `updated_baseline_master_12` will be merged to `master` of the `checkedc-clang` repo. (Note: Certain commits like the commits related to the 3C merge and other commits on the `master` branch will have to be ignored during the review.)

The `updated_baseline_master_12_checkpt` of the `checkedc-clang` repository is at:
  - the branch point of version 12.x on `main`  branch of the upstream LLVM/clang repository with
  - `master` branch (**as of mid-June**) of `checkedc-clang` repo merged into it, plus
  - **fixes for all merge and compilation issues only**.
  
The `updated_baseline_master_12` of the `checkedc-clang` repository is at:
  - the branch point of version 12.x on `main`  branch of the upstream LLVM/clang repository with
  - `master` branch (**current**) of `checkedc-clang` repo merged into it, plus
  - fixes for all merge and compilation issues plus **fixes for all test failures**.

### CODE REVIEW

**Comment by mgrang** [comment]
@sulekhark I reviewed the commits made by you to fix various issues with the merge of LLVM 12.x. The changes look good. Thanks.

**Review by mgrang — APPROVED** [review]

**Comment by mattmccutchen-cci** [comment]
Nice work on the workaround for `clang-tools-extra/clangd/test/checkedc-typevariabletypeloc-uninit.test`.

I assume you've already run the 3C regression tests, but I'll run some additional 3C and clangd tests so that if I find problems, we can discuss whether and how to fix them before the upcoming Checked C release.

**Comment by sulekhark** [comment]
@mattmccutchen-cci , Thanks and yes, I have run `check-3c` and `check-clangd` on the LLVM 12 merge. If you have additional `3C` and/or `clangd` tests, it will be great if you run those. It may be a good idea to run your extra tests on the branch `updated_baseline_master_12_temp`. Here's why:
1. The branch `updated_baseline_master_12` contains the merge of `master` with the upstream LLVM `main` branch up to the commit <branch-point-of-release-12.x-on main>. 
2. The branch `updated_baseline_master_12_temp` is the same as branch `updated_baseline_master_12` but with additional fixes from the upstream  `release/12.x` branch (up to the commit corresponding to 12.0.1 release) merged into it. I have done this in a "temp" branch to ensure that there (hopefully) won't be surprise issues when I later do things in the following sequence:
    -  I will first be merging `updated_baseline_master_12` to `master`.
    - Next, I will create a `release_12.x` branch off  `master`.
    - Then, I will merge all the fixes on the upstream `release/12.x` branch (up to the commit corresponding to 12.0.1 release) on to the `release_12.x` branch just created.

**Comment by mattmccutchen-cci** [comment]
Thanks for the information.  I'll be happy to take your recommendation on which branch(es) to test.  However, I noticed some issues with the branches that you might want to address first, since you'll probably want to address them before the final merge anyway to avoid surprises:

> 2. The branch `updated_baseline_master_12_temp` is the same as branch `updated_baseline_master_12` but with additional fixes from the upstream  `release/12.x` branch (up to the commit corresponding to 12.0.1 release) merged into it.

It seems this isn't quite true: `updated_baseline_master_12` has [a few commits](https://github.com/microsoft/checkedc-clang/compare/updated_baseline_master_12_temp...updated_baseline_master_12) that are not in `updated_baseline_master_12_temp`.

Also, there are [a few commits](https://github.com/microsoft/checkedc-clang/compare/updated_baseline_master_12...master) in `master` but not `updated_baseline_master_12`.  Before I received your comment, I planned to test a temporary merge of `master` and `updated_baseline_master_12`.  When I performed that merge, I got a conflict in `clang/lib/Sema/SemaBounds.cpp` and a compile error in `clang/lib/AST/NormalizeUtils.cpp`, both apparently related to an interaction between #1135 and the `isIntegerConstantExpression` API change that you mostly addressed in 4efec476fc7cc1ff126b24ed532e903e74e09c70.

Aside: I'd suggest that a more meaningful name for `updated_baseline_master_12_temp` would be something like `updated_baseline_master_12_release`, because the difference from `updated_baseline_master_12` is that it includes the commits from the upstream release branch.

File: clang/lib/AST/NormalizeUtils.cpp
Comment (sulekhark) [reply clang/lib/AST/NormalizeUtils.cpp:203]: Thanks for catching this! I will fix this locally and also add a test case.
And, yes, both `updated_baseline_master_12` and `updated_baseline_master_12_temp` are a bit behind `master` (`updated_baseline_master_12_temp` more so).  I plan merge `master` into these branches before proceeding further.

File: clang/lib/AST/NormalizeUtils.cpp
Comment (mattmccutchen-cci) [inline clang/lib/AST/NormalizeUtils.cpp:282]: I see you fixed the same build error that I encountered when I tried to merge `master` and `updated_baseline_master_12`.  But doesn't this need a `getIntegerConstantExpr` now that `isIntegerConstantExpr` no longer does that as a side effect?

```suggestion
  if (BO->getLHS()->getType()->isPointerType() &&
      BO->getRHS()->isIntegerConstantExpr(S.Context)) {
    PointerExpr = BO->getLHS();
    Constant = *BO->getRHS()->getIntegerConstantExpr(S.Context);
  } else if (BO->getOpcode() == BinaryOperatorKind::BO_Add &&
              BO->getRHS()->getType()->isPointerType() &&
              BO->getLHS()->isIntegerConstantExpr(S.Context)) {
    PointerExpr = BO->getRHS();
    Constant = *BO->getLHS()->getIntegerConstantExpr(S.Context);
  } else
```

As with the `NormalizeUtil::GetRHSConstant` bug, if this wasn't caught by the existing Checked C tests, you might want to add more tests.

Comment (sulekhark) [reply clang/lib/AST/NormalizeUtils.cpp:282]: @mattmccutchen-cci , thanks! There were test case failures and I was looking into those.

At this point, all features that were planned for the current release are on `master`. So, now I will work on finalizing the merge and the release (the released Checked C compiler will be based on LLVM 12.0.1). If you are still having merge issues/test failures with `updated_baseline_master_12` or `updated_baseline_master_12_temp`, I propose that you wait until the release, so that you don't have to work with branches that are behind `master`.

**Comment by mattmccutchen-cci** [comment]
I ran CCI's additional 3C and clangd tests on the current `updated_baseline_master_12_temp` branch (b484a29955d357b3ef6da9c68a407e5393ae5ebd) and didn't find any problems.  We could discuss whether we want to run the extra CCI tests on your final release candidates in general, but at least in this case, I think the risk that any of the subsequent Checked C changes on `master` break CCI's tests is pretty low; the important thing was to test the LLVM 12 upgrade.  In the worst case, CCI can offer a branch for anyone who wants the Checked C release plus minimal fixes for any 3C or clangd problems, although they would have to build from source.

**Comment by sulekhark** [comment]
Closing this PR as the changes related to LLVM/Clang 12 upgrade have been merged into the `master` branch.

**Pull request closed without merging**
